llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp

```diff
@@ -2616,6 +2616,20 @@ bool SelectionDAG::expandMultipleResultFPLibCall(
 
   auto [Call, CallChain] = TLI->LowerCallTo(CLI);
 
+  if (CallRetResNo && !Node->hasAnyUseOfValue(*CallRetResNo)) {
+    // FIXME: This is needed for x87, which uses a floating-point stack. If (for
+    // example) the node to be expanded has two results one floating-point which
+    // is returned by the call, and one integer result, returned via an output
+    // pointer. If only the integer result is used then the `CopyFromReg` for
+    // the FP result may be optimized out. This prevents an FP stack pop from
+    // being emitted for it. The `FAKE_USE` node prevents optimizations from
+    // removing the `CopyFromReg` from the chain, and ensures the FP pop will be
+    // emitted. Note: We use an UNDEF pointer as the argument to prevent keeping
+    // any real values live longer than we need to.
+    CallChain = getNode(ISD::FAKE_USE, DL, MVT::Other, CallChain,
+                        getUNDEF(TLI->getPointerTy(getDataLayout())));
```

Comment:
I think an alternative to using `ISD::FAKE_USE` with an undef non-chain operand, would be to define a new node `ISD::CHAIN_BARRIER`, which only takes a chain, which would solely exists to prevent optimizations change the chain from looking passed the "`ISD::CHAIN_BARRIER`" .  I held off on doing that here as `FAKE_USE chain, undef` works too (and I think optimizations should ignore fake_use) , which avoids plumbing the new node.

Reply:
I've gone with adding the new `CHAIN_BARRIER` node, which I think is nicer (no fake_use debug printing, or restricted scheduling) -- and has semantics defined to fix this issue.